src/riak_kv_entropy_manager.erl

```diff
@@ -550,6 +598,66 @@ maybe_poke_tree(State) ->
     riak_kv_index_hashtree:poke(Tree),
     State2.
 
+-spec maybe_start_upgrade(riak_core_ring(),state()) -> state().
+maybe_start_upgrade(Ring, State=#state{trees=Trees, version=undefined, pending_version=undefined}) ->
+    Indices = riak_core_ring:my_indices(Ring),
+    case riak_core_capability:get({riak_kv, object_hash_version}, undefined) of
+        0 when length(Trees) == length(Indices) ->
+            maybe_upgrade(State);
+        _ ->
+            State
+    end;
+maybe_start_upgrade(_Ring, State) ->
+    State.
+
+-spec maybe_upgrade(state()) -> state().
+maybe_upgrade(State=#state{trees_version = VTrees}) ->
+    case [Idx || {Idx, undefined} <- VTrees] of
+        %% Upgrade is done already, set version in state
+        [] ->
+            State#state{version=0};
+        %% No trees have been upgraded, need to wait for exchanges
+        Trees when length(Trees) == length(VTrees) ->
+            check_exchanges_and_upgrade(State);
+        %% Upgrade already started, set pending_version
+        _ ->
+            State#state{pending_version=0}
+    end.
+
+-spec check_exchanges_and_upgrade(state()) -> state().
+check_exchanges_and_upgrade(State) ->
+    case riak_kv_entropy_info:all_sibling_exchanges_complete() of
+        true ->
+            lager:info("Starting AAE hashtree upgrade"),
+            State#state{pending_version=0};
+        false ->
+            State
+    end.
+
+-spec check_upgrade(state()) -> state().
+check_upgrade(State=#state{pending_version=undefined}) ->
+    State;
+check_upgrade(State=#state{pending_version=Version,trees_version=VTrees}) ->
+    %% Verify we have all partitions registered with a hashtree, version
+    {ok, Ring} = riak_core_ring_manager:get_my_ring(),
+    Indices = riak_core_ring:my_indices(Ring),
+    Existing = dict:from_list(VTrees),
+    MissingIdx = [Idx || Idx <- Indices,
+        not dict:is_key(Idx, Existing)],
+    case MissingIdx of
+        [] ->
+            case [Idx || {Idx, V} <- VTrees, V == Version] of
+                [] ->
```

Comment:
I see the `[]` case for when none of the local hashtrees have been upgraded, and then the following case where all local trees have been upgraded, but what if only some have been upgraded? Is that a case we might run into and then end up crashing here?